internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+
+	// Return a copy of an intermediate result.
+	// Will not be able to access final output.
+	// If WaitForCompletion is true, this option is overwritten.
+	ReturnCopy bool `json:"return-copy"`
+
+	// A context that can be used to cancel the job.
+	Context context.Context `json:"-"`
+}
+
+// Job provides system level customization of the job's execution.
+// It supports the following optional functionality:
+//   - synchronization
+//   - context cancellation
+//   - record retrieval
+//   - single record retrieval
+//   - record storage
+//   - record removal from storage
+//   - stat tracking
+//
+// Given job options, input items, and a work function,
+// Job will process a list of input items through the work func in a parallel pool to produce a JobOutput.
+// If a lock key is provided, Job will block on that key before running preventing overuse of resources or conflicts between the same Job.
+// Provide an input key generation function to synchronize at the input item level.
+type Job[InputType any, OutputType any] struct {
+	// The user level options for a Job.
+	*JobOptions
+
+	// The number of workers in the parallel pool.
+	PoolSize int
+
+	// An optional key to lock on.
+	LockKey string
+
+	// A sorted list of items that need work.
+	WorkItems []InputType
+
+	// An optional function to retrieve existing records that should not be processed.
+	// Returns a list of processed records, remaining items, and an error.
+	RetrieveFunc func([]InputType) ([]OutputType, []InputType, error)
+
+	// An optional function to retrieve a record based on a single work item.
+	SingleRetreiveFunc func(InputType) (OutputType, bool, error)
+
+	// An optional function to store the result.
+	StoreFunc func([]OutputType) error
+
+	// An optional function to remove existing records from storage.
+	RemoveStorageFunc func([]InputType) error
+
+	// A function to transform work items into results.
+	// Returns a result, the time of computation, and an error.
+	WorkFunc func(InputType) (OutputType, error)
+
+	// An optional function to get the locking key for an individual work item.
+	WorkItemKeyFunc func(InputType) string
+
+	// An optional function to store stats on how long the job took.
+	StatFunc func(int64)
+}
+
+// JobOutput is the result of a call to Job.Run().
+// It contains a system error, a map of work errors, the result items,
+// the remaining items, the total run time, and a channel that signals execution is complete.
+// If the context is cacelled while running a job,
+// the returned JobOutput will be empty.
+type JobOutput[InputType any, OutputType any] struct {
+	// An error for Job.Run().
+	Error error
+
+	// A map of errors encountered while running the job.
+	// The key of the error will be the index of the item
+	// that failed in RemainingItems.
+	WorkErrors map[int]error
+
+	// The list of results.
+	ResultItems []OutputType
+
+	// The list of tems that still need work.
+	RemainingItems []InputType
+
+	// The total computation time.
```

Comment:
Expand this comment a bit to emphasize that this is only the total time spent in WorkFunc().